api/v1alpha1/gateway_types.go

```diff
@@ -179,56 +182,101 @@ type ListenerTLS struct {
 
 // GatewayStatus defines the observed state of Gateway
 type GatewayStatus struct {
-	// TODO overall status
-
-	// Listeners is the status for each listener block in the
-	// Spec. The status for a given block will match the order as
-	// declared in the Spec, e.g. the status for Spec.Listeners[3]
-	// will be in Status.Listeners[3].
+	// Conditions describe the current conditions of the Gateway.
+	Conditions []GatewayCondition `json:"conditions"`
+	// Listeners provide status for each listener defined in the Spec. The name
+	// in ListenerStatus refers to the corresponding Listener of the same name.
 	Listeners []ListenerStatus `json:"listeners"`
-	// Routes is the status for each attached route to the
-	// Gateway. The status for a given route will match the orer as
-	// declared in the Spec, e.g. the status for Spec.Routes[3] will
-	// be in Status.Routes[3].
-	Routes []GatewayRouteStatus `json:"routes"`
+}
+
+// GatewayConditionType is a type of condition associated with a Gateway.
+type GatewayConditionType string
+
+const (
+	// ConditionNoSuchGatewayClass indicates that the specified GatewayClass
+	// does not exist.
+	ConditionNoSuchGatewayClass GatewayConditionType = "NoSuchGatewayClass"
```

Comment:
For some of these that apply to Gateway itself and not some subset of related things (e.g. Listener, Route) I'm just looking for some help understanding conventions...

It seems to be enumerating possible states which abstractly represent unavailability. How does this approach of representing them as condition _types_ compare to representing them as _reasons_ associated with a general "Unavailable" condition type? One reason might be because the Gateway is unavailable for multiple reasons.

However, even if using multiple condition types is preferable, does there still need to be a higher level condition that represents, in aggregate, "the gateway is unavailable"? Otherwise, which permutation of condition types on the Gateway represents total/general unavailability and how would users know that?

Reply:
I *think* it's that by defining types we are making it clear what the list of conditions should be. The types are just a proxy for defining the `Reason` field anyway. Having the types means that we are constructing a canonical list of Reasons that the API defines. I don't know if this is the best way to do it, but sounds like this is pretty up in the air (from Clayton's comment earlier), so we will be also defining the best practice to an extent in this repo.

How does that interact with 'Use positive conditions that mirror fundamental states of your resource'? I'm not sure.

Reply:
Thanks, Nick. To be clear, it sounds like this doesn't translate into any specific requirement/contract for implementations (i.e. "implementations _must_ publish [condition set]"), but they _should_ implement them consistently?

Reply:
I think so.